File: praticaUmModularizacao/qumbsk.c
/*
receber a b e c
se a = 0 sair
calcular_raizes func()
a maior que 0 e menor que 10.0
b <= -1000.0
c <= 1000.0


*/
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
void preencher(float *a, float *b, float *c)
{
    float temp=0.0;
    char var[] = {'A','B','C'};

    printf("\naquisicao de valores\n");
    for(int i=0; i<=2; i++)
    {
     printf("insira o valor de %c: ", var[i]);
     scanf("%f", &temp);
        if(i==0) //variavel a
        {  
            if(temp==0)
            {
                printf("\n*****nao pode ser 0******");
                temp=0.0;
                i--;
                continue;
            }
            *a=temp;
                
        }else{
            if(i>=1)
            {
                if(temp <-1000.0 || temp>1000.0)
                {   
                    printf("\n*****fora do intervalo******");//refazzer
                    i--;
                    continue;
                }
                if(i==1)
                {
                    *b=temp;
                }
                if(i==2)
                {
                    *c=temp;
                }
            }
        }
    }
}
float raiz_quadrada(float numero)
{
    float tolerancia = 0.0000001;
    float chute = numero/2.0;
    float proximo_chute;
    while(1)
    {
        proximo_chute =0.5*(chute+(numero/chute));
        if ((chute - proximo_chute) < tolerancia && (chute - proximo_chute) > -tolerancia)
        {
            break;
        }
        chute=proximo_chute;
    }
return chute;
}
float disciminante(float *a, float *b, float *c)
{
    float disc=0;
    disc=((pow(*b, 2))-(4* (*a * *c))); //necessario "*" na frente da variavel pra usar
    if(disc <0)
    {
     return -1;   //impossivel calcular e sair
    }
return disc;
}
void calcular_raizes (float *a, float *b, float *c, float *x1, float *x2)
{

    float discriminante = 0.0;
    if((discriminante =disciminante(a,b,c)) ==-1)
    {
        printf("\nimpossivel calcular ");
        exit(1);
    }else{
        float raiz_discriminante = raiz_quadrada(discriminante);
        *x1=(((-*b)+raiz_discriminante)/(2**a));
        *x2=(((-*b)-raiz_discriminante)/(2**a));
    }

}
int main(int argc, char *argv[] )
{

    float *a, *b, *c, *x1, *x2;
    a=(float*) calloc(1, (sizeof(float)));
    b = (float*)calloc (1,(sizeof(float)));
    c = (float*)calloc(1,(sizeof(float)));
    x1 = (float*) calloc(1,sizeof(float));
    x2=(float*)calloc(1,sizeof(float));

    preencher(a, b, c);
    calcular_raizes(a, b, c, x1, x2);
    printf("\nvalor de x1: %f \nvalor de x2: %f", *x1, *x2);

return 0;
}
/*
questao 1 modularização

*/

File: praticaUmModularizacao/qumbsk.lua
--[[
comentarios 

receber a b e c
se a = 0 sair
calcular_raizes func()
a maior que 0 e menor que 10.0
b <= -1000.0
c <= 1000.0

\/
raizQuadrada();//calcular raiz
discriminantes();//calcular o discriminante
obterRAizes();//calcular raiz x1 e x2
obeterValores();//obter a,b e c
main();//main

__index = ação padrao quando algo nao ta deifinido na metatabela
setmetatable=concatena {[tabela], [outra tabela]}
--]]
local function raiz_quadrada(numero)
    local tolerancia =0.0000001;
    local chute = numero/2.0;
    local proximo_chute=0.0;
        while 1 do
            proximo_chute=0.5*(chute+(numero/chute));
            if ((chute - proximo_chute) < tolerancia and (chute - proximo_chute) > -tolerancia) then
                break;
            end
            chute=proximo_chute;
        end
    return chute;
end
local function discriminante(var)
    local retorno =0.0;
    retorno=((math.pow(var.B, 2))-(4 * (var.A * var.C)));
    if retorno<0 then
        --sair pos da erro
        error("discriminante nao pode ser negativo", 1)
    end
    return retorno;
end
    
local function obterRAizes(var, raizes) 

        --chamar calcular discriminante
        local varDiscriminante = 0.0;
        varDiscriminante=discriminante(var);

        --if
        if(varDiscriminante==-1) then
            error("impossivel calcular: disciminante -1",2)
  
        else
            local raiz_discriminante=raiz_quadrada(varDiscriminante);--raiz do discminante 
            raizes.Um = ((((-var.B)+raiz_discriminante)/(2*var.A))); --raiz um +
            raizes.Dois = ((((-var.B)-raiz_discriminante)/(2*var.A))); --raiz dois -
        end
end

local function obterValores(var)--passgaem de tabelas é referencia 
        while 1 do
            print("***inserir valores***")
                while 1 do
                    --inserir A
                    ::continue::
                    io.write("\ninerir A: ");
                    local varTemp=io.read("*n");
                        if varTemp<=0 then --se A for menor que 0
                            goto continue
                        else
                            if varTemp>1 then
                                --atribuir a var.A
                                var.A=varTemp;
                                break;
                            end
                        end
                end
                while 1 do
                    --inserir b
                    ::continuedois::
                    io.write("\ninserir B: ");
                    local varTemp=io.read("*n");
                        if varTemp <=-1000.0 or varTemp>=1000.1 then
                            -- mandar pro inico de inserir b 
                            goto continuedois
                        else 
                                --atribuir a var.b
                                var.B=varTemp;
                            break;
                        end 
                end
                while 1 do
                    --inserir C
                    ::continuetres::
                    io.write("\ninserir C: ");
                    local varTemp=io.read("*n");
                        if varTemp<=-1000.0 or varTemp>=1000.1 then
                            --mandar pro inicio de inserir B
                            goto continuetres;
                        else
                            var.C=varTemp;
                            break;
                        end
                end
                break;
        end
end
    

local function main()
--ler a,b e c
    local var={A=0.0, B=0.0, C =0.0} --unica estrutura é tabela
    local raizes={Um=0.0, Dois=0.0}
    --io.wrire()
    --=io.read()
    obterValores(var);
    obterRAizes(var, raizes);
    print("\n***saida:***");
    io.write("\nraiz um: "..raizes.Um);
    io.write("\nraiz Dois: "..raizes.Dois.."\n");
end
main();